net/src/memory_book.rs

```diff
@@ -0,0 +1,153 @@
+// Books
+//--------------------------------------------------------------------------------------------------
+
+/// Type for holding list of metaId per entry
+/// i.e. map of entry_address -> Set(metaId)
+/// edge case is entry_address -> entry_address, which is the way to signal that we are holding
+/// the entry itself
+pub(crate) type EntryBook = HashMap<Address, HashSet<HashString>>;
+
+/// Type for holding list of addresses per entry per dna+agent_id
+/// i.e. map of cell_id -> EntryBook
+pub(crate) type CellBook = HashMap<CellId, EntryBook>;
+
+/// Add an address to a book
+pub(crate) fn bookkeep_with_cell_id(
+    cell_book: &mut CellBook,
+    cell_id: CellId,
+    base_address: &Address,
+    data_address: &Address,
+) {
+    // Append to existing address list if there is one
+    {
+        let maybe_entry_book = cell_book.get_mut(&cell_id);
+        if let Some(entry_book) = maybe_entry_book {
+            // Append to existing address list if there is one
+            {
+                let maybe_meta_set = entry_book.get_mut(&base_address);
+                if let Some(meta_set) = maybe_meta_set {
+                    meta_set.insert(data_address.clone());
+                    return;
+                }
+            }
+            let mut meta_set = HashSet::new();
+            meta_set.insert(data_address.clone());
+            entry_book.insert(base_address.clone(), meta_set.clone());
+            return;
+        }
+    } // unborrow book
+      // None: Create and add a new EntryBook
+    let mut entry_book = EntryBook::new();
+    let mut meta_set = HashSet::new();
+    meta_set.insert(data_address.clone());
+    entry_book.insert(base_address.clone(), meta_set);
+    cell_book.insert(cell_id, entry_book);
+}
+
+/// Add an address to a book (sugar)
+pub(crate) fn bookkeep(
+    cell_book: &mut CellBook,
+    dna_address: &Address,
+    agent_id: &str,
+    base_address: &Address,
+    data_address: &Address,
+) {
+    let cell_id = into_cell_id(dna_address, agent_id);
+    bookkeep_with_cell_id(cell_book, cell_id, base_address, data_address);
+}
+
+// Return true if data is in book
+pub fn book_has(
+    cell_book: &CellBook,
+    cell_id: CellId,
+    base_address: &Address,
+    data_address: &Address,
+) -> bool {
+    let maybe_entry_book = cell_book.get(&cell_id);
+    if let None = maybe_entry_book {
+        return false;
+    }
+    let entry_book = maybe_entry_book.unwrap();
+    let maybe_meta_set = entry_book.get(base_address);
+    if let None = maybe_meta_set {
+        return false;
+    }
+    let meta_set = maybe_meta_set.unwrap();
+    meta_set.contains(data_address)
+}
+
+///
+pub fn book_has_entry(cell_book: &CellBook, cell_id: CellId, entry_address: &Address) -> bool {
+    book_has(&cell_book, cell_id, entry_address, entry_address)
+}
+
+/// Remove an address from a book
+/// Return true if address exists and has been successfully removed.
+pub(crate) fn _unbookkeep_address(
```

Comment:
So, there is absolutely no convention in Rust for functions with underscores as a prefix. We do not implement this convention anywhere else in the codebase. Please remove it.

Reply:
Isn't that the convention for unused variables/functions? This function is unused, so the underscore omits the warning, right? So the question is, if it is unused, can we remove it?

Reply:
Yup. If it's unused, we should remove it

Reply:
My point it, that you shouldn't omit the warning but act upon it

Reply:
I'm going to need those function in the future. Not going to remove them.